### PR TITLE
Feature 179 flann

### LINKED ISSUE #179: Add FLANN based matching
Add FLANN (Fast Library for Approximate Nearest Neighbors) matching to the repo. This should be faster than BruteForceMatching, hopefully bringing feature matching closer to realtime operation.

### PR DESCRIPTION
**Pull request by navganti**
Resolves #179. Adds FLANN based matching to libwave. FLANN is the Fast Library for Approximate Nearest Neighbours, which has a variety of different methods to separate objects (descriptors in our case) into various buckets for quick lookup. There are extensive configuration parameters here, and so in the interests of time I have just implemented the defaults for a few of the methods. This will be expanded upon in some point in the future.

#### Pre-Merge Checklist:
- [x] Code is documented in doxygen format
- [x] Code has automated tests
- [x] Zero compiler warnings
- [x] Formatted with `clang-format`

### CODE REVIEW

File: wave_vision/src/matcher/flann_matcher.cpp [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+FLANNMatcher::FLANNMatcher(const FLANNMatcherParams &config) {
+    // Check flann_method and create the appropriate parameters struct.
+    this->checkConfiguration(config);
+
+    // Depending on the FLANN method, different parameters are required.
+    if (config.flann_method == FLANN::KDTree) {
+        // Create FLANN matcher with default KDTree and Search params.
+        cv::FlannBasedMatcher matcher(new cv::flann::KDTreeIndexParams(),
```
Comment (leokoppel) [inline wave_vision/src/matcher/flann_matcher.cpp]: Avoiding `new` is preferable (with `cv::makePtr`?) but I _think_ in this case opencv will take ownership of the pointer anyway.

File: wave_vision/src/matcher/flann_matcher.cpp [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    if (check_config.fm_method != cv::FM_7POINT &&
+        check_config.fm_method != cv::FM_8POINT &&
+        check_config.fm_method != cv::FM_LMEDS &&
+        check_config.fm_method != cv::FM_RANSAC) {
+        throw std::invalid_argument("fm_method is not an acceptable value!");
+    }
+}
+
+std::vector<cv::DMatch> FLANNMatcher::filterMatches(
+  std::vector<cv::DMatch> &matches) const {
```
Comment (leokoppel) [inline wave_vision/src/matcher/flann_matcher.cpp]: should this param be const?

File: wave_vision/src/matcher/flann_matcher.cpp (hunk 1 above)
Comment (navganti) [inline wave_vision/src/matcher/flann_matcher.cpp]: done

File: wave_vision/src/matcher/flann_matcher.cpp (hunk 2 above)
Comment (navganti) [inline wave_vision/src/matcher/flann_matcher.cpp]: done

**Pull request merged**